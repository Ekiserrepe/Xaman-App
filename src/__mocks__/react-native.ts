/* eslint-disable */

const ReactNative = require('react-native');
const crypto = require('crypto');

ReactNative.NativeModules.CryptoModule = {
    randomKeySync: jest.fn((len: number) => crypto.randomBytes(len).toString('hex').toUpperCase()),
    randomKey: jest.fn((len: number) => Promise.resolve(crypto.randomBytes(len).toString('hex').toUpperCase())),
    sha512: jest.fn((value) => Promise.resolve(crypto.createHash('sha512').update(value).digest('hex'))),
    sha256: jest.fn((value) => Promise.resolve(crypto.createHash('sha256').update(value).digest('hex'))),
    sha1: jest.fn((value) => Promise.resolve(crypto.createHash('sha1').update(value).digest('hex'))),
    hmac256: jest.fn((value, key) =>
        Promise.resolve(crypto.createHmac('sha256', Buffer.from(key, 'hex')).update(value).digest('hex')),
    ),
};

ReactNative.NativeModules.UtilsModule = {
    flagSecure: jest.fn((enable: boolean) => true),
    isRooted: jest.fn(() => false),
    isJailBroken: jest.fn(() => false),
    getTimeZone: jest.fn(() => 'Europe/Amsterdam'),
    hapticFeedback: jest.fn((type: any) => true),
    restartBundle: jest.fn(() => true),
    timeoutEvent: jest.fn((event: string, timeout: number) => true),
    getElapsedRealtime: jest.fn(() => Promise.resolve('1337')),
    exitApp: jest.fn(() => true),
};

ReactNative.NativeModules.LocalNotificationModule = {
    setBadge: jest.fn((badge: number) => Promise.resolve()),
    getBadge: jest.fn(() => Promise.resolve(1)),
    complete: jest.fn((messageId: string, show: boolean) => true),
};

ReactNative.NativeModules.AppUpdateModule = {
    checkUpdate: jest.fn(() => Promise.resolve(123)),
    startUpdate: jest.fn(() => Promise.resolve()),
};

ReactNative.NativeModules.SharedPreferencesModule = {
    get: jest.fn((key) => Promise.resolve('value')),
    set: jest.fn((key, value) => Promise.resolve(true)),
    del: jest.fn((key) => Promise.resolve(true)),
};

ReactNative.NativeModules.Toast = {
    showWithGravity: jest.fn((message: string, duration: any, gravity: number) => true),
};

ReactNative.NativeModules.DeviceUtilsModule = {
    layoutInsets: { top: 0, bottom: 0 },
    brand: 'Apple',
    model: 'iPhone13,4',
    osVersion: '15,5',
    getElapsedRealtime: jest.fn(() => Promise.resolve(0)),
    isJailBroken: jest.fn(() => Promise.resolve(false)),
    isRooted: jest.fn(() => Promise.resolve(false)),
    getTimeZone: jest.fn(() => Promise.resolve('Europe/Amsterdam')),
    getLocalSetting: jest.fn(() =>
        Promise.resolve({ delimiter: ',', languageCode: 'en', locale: 'en_US', separator: '.' }),
    ),
};

ReactNative.NativeModules.AppUtilsModule = {
    appVersion: '0.0.1',
    buildNumber: '1',
    isDebug: false,
    isFlagSecure: jest.fn(() => Promise.resolve(false)),
    setFlagSecure: jest.fn((value) => {}),
    restartBundle: jest.fn(() => {}),
    exitApp: jest.fn(() => {}),
    timeoutEvent: jest.fn((key, delay) => {}),
};

ReactNative.NativeModules.UniqueIdProviderModule = {
    getDeviceUniqueId: jest.fn(() => 'e988b7a9-f685-4674-87bc-0ad52a52faa5'),
};

ReactNative.NativeModules.HapticFeedbackModule = {
    trigger: jest.fn((type) => {}),
};

ReactNative.NativeModules.VaultManagerModule = {
    latestCipherVersion: 2,
    getStorageEncryptionKey: jest.fn(() =>
        Promise.resolve(
            '1567F58A794600717029077C34A8FAAB9B16B9FFAB174248DD296DA82084EE7921E51DC5757CA655271AF4928263FEC4A36D2139AD02F9CB1BC70F8FD7D38796',
        ),
    ),
    isStorageEncryptionKeyExist: jest.fn(() => Promise.resolve(true)),
    createVault: jest.fn((vaultName: string, entry: string, key: string) => Promise.resolve(true)),
    openVault: jest.fn((vaultName: string, key: string) => Promise.resolve('clearText')),
    vaultExist: jest.fn((vaultName: string) => Promise.resolve(true)),
    purgeVault: jest.fn((vaultName: string) => Promise.resolve(true)),
    reKeyVault: jest.fn((vaultName: string, oldKey: string, newKey: string) => Promise.resolve(true)),
    reKeyBatchVaults: jest.fn((vaultNames: string[], oldKey: string, newKey: string) => Promise.resolve(true)),
    clearStorage: jest.fn(() => Promise.resolve(true)),
    isMigrationRequired: jest.fn((vaultName: string) =>
        Promise.resolve({
            vault: vaultName,
            current_cipher_version: 2,
            latest_cipher_version: 2,
            migration_required: false,
        }),
    ),
};

ReactNative.NativeModules.InAppPurchaseModule = {
    isUserPurchasing: jest.fn((type) => false),
};

ReactNative.Animated.timing = () => ({
    start: (cb?: () => void) => (cb ? cb() : undefined),
});

module.exports = ReactNative;
