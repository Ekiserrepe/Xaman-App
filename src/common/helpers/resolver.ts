/**
 * AccountResolver is responsible for resolving account names and retrieving account information.
 * It provides utility methods to look up account names based on the address and tag,
 * as well as methods to fetch detailed account information including risk level and settings.
 */

import { has, get, assign } from 'lodash';

import AccountRepository from '@store/repositories/account';
import ContactRepository from '@store/repositories/contact';

import LedgerService from '@services/LedgerService';
import BackendService from '@services/BackendService';

import Amount from '@common/libs/ledger/parser/common/amount';

import LRUCache from '@common/utils/cache';
import LoggerService, { LoggerInstance } from '@services/LoggerService';

/* Types ==================================================================== */
export interface PayIDInfo {
    account: string;
    tag: string | null;
}

export interface AccountNameType {
    address: string;
    tag?: number;
    name?: string;
    source?: string;
    kycApproved?: boolean;
}

export interface AccountInfoType {
    exist: boolean;
    risk: 'ERROR' | 'UNKNOWN' | 'PROBABLE' | 'HIGH_PROBABILITY' | 'CONFIRMED';
    requireDestinationTag: boolean;
    possibleExchange: boolean;
    disallowIncomingXRP: boolean;
    blackHole: boolean;
}

/* Resolver ==================================================================== */
class AccountResolver {
    private static CacheSize = 300;
    private cache: LRUCache<string, Promise<AccountNameType> | AccountNameType>;
    private logger: LoggerInstance;

    constructor() {
        this.cache = new LRUCache<string, Promise<AccountNameType> | AccountNameType>(AccountResolver.CacheSize);
        this.logger = LoggerService.createLogger('AccountResolver');
    }

    private lookupresolveAccountName = async (
        address: string,
        tag?: number,
        internal = false,
    ): Promise<AccountNameType> => {
        const notFound: AccountNameType = {
            address,
            tag,
            name: '',
            source: '',
        };

        if (!address) {
            return notFound;
        }

        // Check in address book
        try {
            const contact = await ContactRepository.findOne({
                address,
                destinationTag: `${tag ?? ''}`,
            });

            if (contact) {
                return {
                    address,
                    tag,
                    name: contact.name,
                    source: 'contacts',
                };
            }
        } catch (error) {
            this.logger.error('fetching contact:', error);
        }

        // Check in accounts list
        try {
            const account = await AccountRepository.findOne({ address });
            if (account) {
                return {
                    address,
                    tag,
                    name: account.label,
                    source: 'accounts',
                };
            }
        } catch (error) {
            this.logger.error('fetching account:', error);
        }

        // Only lookup for local results
        if (internal) {
            return notFound;
        }

        // Check the backend
        try {
            const res = await BackendService.getAddressInfo(address);
            if (res) {
                return {
                    address,
                    tag,
                    name: res.name ?? undefined,
                    source: res.source?.replace('internal:', '').replace('.com', ''),
                    kycApproved: res.kycApproved,
                };
            }
        } catch (error) {
            this.logger.error('fetching info from API', error);
        }

        return notFound;
    };

    public setCache = (key: string, value: AccountNameType | Promise<AccountNameType>) => {
        this.cache.set(key, value);
    };

    public getAccountName = async (address: string, tag?: number, internal = false): Promise<AccountNameType> => {
        if (!address) {
            throw new Error('Address is required.');
        }

        const key = `${address}${tag ?? ''}`;

        const cachedValue = this.cache.get(key);
        if (cachedValue) {
            return cachedValue;
        }

        const resultPromise = (async () => {
            const result = await this.lookupresolveAccountName(address, tag, internal);
            this.cache.set(key, result);
            return result;
        })();

        this.cache.set(key, resultPromise); // save the promise itself for subsequent calls

        return resultPromise;
    };

    getAccountInfo = async (address: string): Promise<AccountInfoType> => {
        if (!address) {
            throw new Error('Address is required.');
        }

        const info: AccountInfoType = {
            exist: true,
            risk: 'UNKNOWN',
            requireDestinationTag: false,
            possibleExchange: false,
            disallowIncomingXRP: false,
            blackHole: false,
        };

        // get account risk level
        const accountAdvisory = await BackendService.getAccountAdvisory(address);

        if (has(accountAdvisory, 'danger')) {
            assign(info, { risk: accountAdvisory.danger });
        } else {
            this.logger.error('account advisory risk level not found.');
            throw new Error('Account advisory risk level not found.');
        }

        const accountInfo = await LedgerService.getAccountInfo(address);

        // account doesn't exist, no need to check account risk
        if ('error' in accountInfo) {
            if (get(accountInfo, 'error') === 'actNotFound') {
                assign(info, { exist: false });
                return info;
            }
            this.logger.error('fetching account info:', accountInfo);
            throw new Error('Error fetching account info.');
        }

        const { account_data, account_flags } = accountInfo;

        // if balance is more than 1m possibly exchange account
        if (has(account_data, ['Balance'])) {
            if (new Amount(account_data.Balance, true).dropsToNative().toNumber() > 1000000) {
                assign(info, { possibleExchange: true });
            }
        }

        // check for black hole
        if (has(account_data, ['RegularKey'])) {
            if (
                account_flags?.disableMasterKey &&
                ['rrrrrrrrrrrrrrrrrrrrrhoLvTp', 'rrrrrrrrrrrrrrrrrrrrBZbvji'].indexOf(account_data.RegularKey ?? '') >
                    -1
            ) {
                assign(info, { blackHole: true });
            }
        }

        // check for disallow incoming XRP
        if (account_flags?.disallowIncomingXRP) {
            assign(info, { disallowIncomingXRP: true });
        }

        if (get(accountAdvisory, 'force_dtag')) {
            // first check on account advisory
            assign(info, { requireDestinationTag: true, possibleExchange: true });
        } else if (account_flags?.requireDestinationTag) {
            // check if account have the required destination tag flag set
            assign(info, { requireDestinationTag: true, possibleExchange: true });
        } else {
            // scan the most recent transactions of the account for the destination tags
            const transactionsResp = await LedgerService.getTransactions(address, undefined, 200);
            if (
                !('error' in transactionsResp) &&
                transactionsResp.transactions &&
                transactionsResp.transactions.length > 0
            ) {
                const incomingTXS = transactionsResp.transactions.filter((tx) => tx.tx.Destination === address);

                const incomingTxCountWithTag = incomingTXS.filter(
                    (tx) =>
                        typeof tx.tx.TransactionType === 'string' &&
                        typeof tx.tx.DestinationTag !== 'undefined' &&
                        Number(tx.tx.DestinationTag) > 9999,
                ).length;

                const senders = transactionsResp.transactions.map((tx) => tx.tx.Account || '');

                const uniqueSenders = new Set(senders).size;

                const percentageTag = (incomingTxCountWithTag / incomingTXS.length) * 100;

                if (uniqueSenders >= 10 && percentageTag > 50) {
                    assign(info, { requireDestinationTag: true, possibleExchange: true });
                }
            }
        }

        return info;
    };

    getPayIdInfo = (payId: string): Promise<PayIDInfo | undefined> => {
        return BackendService.lookup(payId)
            .then((res) => {
                if (res) {
                    if (Array.isArray(res.matches) && res.matches.length > 0) {
                        const match = res.matches[0];
                        return {
                            account: match.account,
                            tag: match.tag,
                        };
                    }
                }
                return undefined;
            })
            .catch(() => {
                return undefined;
            });
    };
}

export default new AccountResolver();
