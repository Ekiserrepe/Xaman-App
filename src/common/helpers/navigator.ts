import { get } from 'lodash';
import { Platform, InteractionManager } from 'react-native';

import { Navigation, Options, LayoutTabsChildren } from 'react-native-navigation';

import { GetBottomTabScale, HasBottomNotch } from '@common/helpers/device';

import { AppScreens } from '@common/constants';

import Localize from '@locale';

import NavigationService, { ComponentTypes, RootType } from '@services/NavigationService';
import StyleService from '@services/StyleService';

import { Props as AlertOverlayProps } from '@screens/Overlay/Alert/types';

import AppFonts from '@theme/fonts';

/* Types ==================================================================== */
export type AppScreenKeys<T = typeof AppScreens, L0 = T[keyof T]> =
    L0 extends Record<string, any> ? AppScreenKeys<L0> : L0;

type EnforcedProps<P extends { [K in keyof P]: any }> = P;

/* Constants ==================================================================== */
const getDefaultOptions = (): Options => {
    return {
        layout: {
            backgroundColor: StyleService.value('$background'),
            componentBackgroundColor: StyleService.value('$background'),
            orientation: ['portrait'],
            adjustResize: false,
        },
        topBar: {
            visible: false,
        },
        navigationBar: {
            backgroundColor: StyleService.value('$tint'),
        },
        statusBar: {
            style: Platform.select({
                android: undefined,
                ios: StyleService.isDarkMode() ? 'light' : 'dark',
            }),
            drawBehind: false,
        },
        bottomTabs: {
            backgroundColor: StyleService.value('$background'),
            translucent: false,
            animate: false,
            drawBehind: false,
            tabsAttachMode: 'onSwitchToTab',
            titleDisplayMode: 'alwaysShow',
            elevation: 10,
            hideShadow: Platform.OS === 'android',
            shadow: Platform.select({
                android: undefined,
                ios: {
                    opacity: StyleService.isDarkMode() ? 0.13 : 0.07,
                    color: StyleService.isDarkMode() ? 'white' : 'black',
                    radius: StyleService.isDarkMode() ? 12 : 8,
                },
            }),
        },

        animations: {
            pop: {
                enabled: Platform.OS === 'ios',
            },
        },
        popGesture: true,
        blurOnUnmount: true,
    };
};

const getTabBarIcons = (): {
    [k in string]: {
        icon: { uri: string };
        iconSelected: { uri: string };
        scale: number;
    };
} => {
    return {
        [AppScreens.TabBar.Home]: {
            icon: StyleService.getImage('IconTabBarHome'),
            iconSelected: StyleService.getImage('IconTabBarHomeSelected'),
            scale: GetBottomTabScale(0.9),
        },
        [AppScreens.TabBar.Events]: {
            icon: StyleService.getImage('IconTabBarEvents'),
            iconSelected: StyleService.getImage('IconTabBarEventsSelected'),
            scale: GetBottomTabScale(0.9),
        },
        [AppScreens.TabBar.Actions]: {
            icon: StyleService.getImage('IconTabBarActions'),
            iconSelected: StyleService.getImage('IconTabBarActions'),
            scale: GetBottomTabScale(0.65),
        },
        [AppScreens.TabBar.XApps]: {
            icon: StyleService.getImage('IconTabBarXapp'),
            iconSelected: StyleService.getImage('IconTabBarXappSelected'),
            scale: GetBottomTabScale(0.9),
        },
        [AppScreens.TabBar.Settings]: {
            icon: StyleService.getImage('IconTabBarSettings'),
            iconSelected: StyleService.getImage('IconTabBarSettingsSelected'),
            scale: GetBottomTabScale(0.9),
        },
    };
};

/* Lib ==================================================================== */

const Navigator = {
    /**
     * Initializes and sets the default configuration for the navigation tabs.
     * This method starts the app in bottom tabs mode
     *
     * @return {void}
     */
    startDefault(): void {
        const defaultOptions = getDefaultOptions();
        Navigation.setDefaultOptions(defaultOptions);

        const bottomTabStyles = StyleService.applyTheme({
            textColor: '$grey',
            selectedTextColor: '$textPrimary',
            fontFamily: AppFonts.base.familyExtraBold,
            iconInsets: {
                top: HasBottomNotch() ? 4 : 1,
            },
        });

        const TabBarIcons = getTabBarIcons();

        const bottomTabsChildren: LayoutTabsChildren[] = [];

        Object.keys(AppScreens.TabBar).forEach((tab) => {
            bottomTabsChildren.push({
                stack: {
                    id: `bottomTab-${tab}`,
                    children: [
                        {
                            component: {
                                name: tab === 'Actions' ? AppScreens.Global.Placeholder : get(AppScreens.TabBar, tab),
                                id: get(AppScreens.TabBar, tab),
                            },
                        },
                    ],
                    options: {
                        topBar: {
                            visible: false,
                        },
                        bottomTab: {
                            selectTabOnPress: tab !== 'Actions',
                            text: tab !== 'Actions' ? Localize.t(`global.${tab.toLowerCase()}`) : '',
                            icon: {
                                scale: TabBarIcons[get(AppScreens.TabBar, tab)].scale,
                                ...TabBarIcons[get(AppScreens.TabBar, tab)].icon,
                            },
                            selectedIcon: {
                                scale: TabBarIcons[get(AppScreens.TabBar, tab)].scale,
                                ...TabBarIcons[get(AppScreens.TabBar, tab)].iconSelected,
                            },
                            testID: `tab-${tab}`,
                            ...bottomTabStyles,
                        },
                    },
                },
            });
        });

        InteractionManager.runAfterInteractions(() => {
            Navigation.setRoot({
                root: {
                    bottomTabs: {
                        id: RootType.DefaultRoot,
                        children: bottomTabsChildren,
                    },
                },
            });
        });
    },

    /**
     * Starts the app navigation in onboarding screen
     *
     * @return {void}
     */
    startOnboarding(): void {
        const defaultOptions = getDefaultOptions();
        Navigation.setDefaultOptions(defaultOptions);

        InteractionManager.runAfterInteractions(() => {
            Navigation.setRoot({
                root: {
                    stack: {
                        id: AppScreens.Onboarding,
                        children: [
                            {
                                component: {
                                    name: AppScreens.Onboarding,
                                },
                            },
                        ],
                    },
                },
            });
        });
    },

    /**
     * Pushes a new screen to the navigation stack.
     *
     * @template P - Props type
     * @param nextScreen - The key of the screen to push.
     * @param passProps - The props to pass to the pushed screen.
     * @param options - The options for the pushed screen (default: {}).
     * @returns Screen name if the screen was pushed, otherwise false.
     */
    push<P extends object>(
        nextScreen: AppScreenKeys,
        passProps: EnforcedProps<P>,
        options: Options = {},
    ): Promise<string | boolean> {
        const currentScreen = NavigationService.getCurrentScreen() ?? '';
        if (currentScreen !== nextScreen) {
            return Navigation.push(currentScreen, {
                component: {
                    name: nextScreen,
                    id: nextScreen,
                    passProps: Object.assign(passProps, { componentType: ComponentTypes.Screen }),
                    options,
                },
            });
        }

        return Promise.resolve(false);
    },

    /**
     * Pops the current screen from the navigation stack.
     *
     * @param {Options} options - Optional configuration options for the pop operation.
     * @return {Promise} - A promise that resolves when the pop operation is complete.
     */
    pop(options: Options = {}): Promise<string | boolean> {
        const currentScreen = NavigationService.getCurrentScreen();

        if (currentScreen) {
            return Navigation.pop(currentScreen, options);
        }

        return Promise.resolve(false);
    },

    /**
     * Pops to the root screen of the navigation stack.
     *
     * @param {Options} [options={}] - The options for popping to the root screen. (optional)
     * @return {Promise<string | boolean>} - A promise that resolves with no value.
     */
    popToRoot(options: Options = {}): Promise<string | boolean> {
        const currentScreen = NavigationService.getCurrentScreen();

        if (currentScreen) {
            return Navigation.popToRoot(currentScreen, options);
        }

        return Promise.resolve(false);
    },

    /**
     * Displays the specified overlay screen.
     *
     * @template P - Type for the passProps argument
     * @param {AppScreenKeys} overlay - The name of the overlay screen to display.
     * @param {EnforcedProps<P>} passProps - The properties to pass to the overlay screen.
     * @param {Options} [options={}] - Additional options for customizing the overlay's appearance and behavior.
     * @returns {Promise<string | boolean>} - A Promise that resolves to either the overlay's id or false
     */
    showOverlay<P extends object>(
        overlay: AppScreenKeys,
        passProps: EnforcedProps<P>,
        options: Options = {},
    ): Promise<string | boolean> {
        const currentOverlay = NavigationService.getCurrentOverlay();

        if (currentOverlay !== overlay) {
            return Navigation.showOverlay({
                component: {
                    name: overlay,
                    id: overlay,
                    passProps: Object.assign(passProps, { componentType: ComponentTypes.Overlay }),
                    options: {
                        overlay: {
                            handleKeyboardEvents: true,
                        },
                        layout: {
                            backgroundColor: 'transparent',
                            componentBackgroundColor: 'transparent',
                        },
                        ...options,
                    },
                },
            });
        }

        return Promise.resolve(false);
    },

    /**
     * Dismisses the specified overlay or current overlay.
     *
     * @param {AppScreenKeys} [overlay] - The overlay to dismiss. If not provided, the current overlay will be used.
     * @return {Promise<string | boolean>} - A promise that resolves to a string or boolean value.
     */
    dismissOverlay(overlay?: AppScreenKeys): Promise<string | boolean> {
        const overlayToDismiss = overlay ?? NavigationService.getCurrentOverlay();

        if (overlayToDismiss) {
            return Navigation.dismissOverlay(overlayToDismiss);
        }

        return Promise.resolve(false);
    },

    /**
     * Display a modal screen using React Navigation.
     *
     * @template P - Type parameter representing the props object for the modal screen.
     * @param {string} modal - The key name of the modal screen component to be displayed.
     * @param {P} passProps - The props object to be passed to the modal screen component.
     * @param {object} [options={}] - Additional options for the modal screen component.
     * @returns {boolean} - Returns `string` if the modal screen is shown successfully, otherwise `false`.
     */
    showModal<P extends object>(
        modal: AppScreenKeys,
        passProps: EnforcedProps<P>,
        options: Options = {},
    ): Promise<string | boolean> {
        const currentScreen = NavigationService.getCurrentModal();
        if (currentScreen !== modal) {
            return Navigation.showModal({
                stack: {
                    id: modal,
                    children: [
                        {
                            component: {
                                name: modal,
                                id: modal,
                                options,
                                passProps: Object.assign(passProps, { componentType: ComponentTypes.Modal }),
                            },
                        },
                    ],
                },
            });
        }

        return Promise.resolve(false);
    },

    /**
     * Dismisses the current modal.
     *
     * @return {Promise<string | boolean>} A promise that resolves to either string or false
     */
    dismissModal(): Promise<string | boolean> {
        const currentModal = NavigationService.getCurrentModal();

        if (currentModal) {
            return Navigation.dismissModal(currentModal);
        }

        return Promise.resolve(false);
    },

    /**
     * Sets a badge to a specified tab.
     *
     * @param {string} tab - The tab to set the badge on.
     * @param {string} badge - The badge value to set.
     * @return {void}
     */
    setBadge(tab: string, badge: string): void {
        Navigation.mergeOptions(tab, {
            bottomTab: {
                badge,
            },
        });
    },

    /**
     * Displays a modal overlay with an alert message and passed props.
     *
     * @param {AlertOverlayProps} props - The props for the alert overlay.
     * @return {Promsie<string | boolean>}
     */
    showAlertModal(props: AlertOverlayProps): Promise<string | boolean> {
        return Navigator.showOverlay<AlertOverlayProps>(AppScreens.Overlay.Alert, props);
    },

    /**
     * Merges the provided options object with the options of the current screen.
     * If screen is not provided, the current screen is determined using NavigationService.
     *
     * @param {AppScreenKeys} screen - The screen for which to merge the options. defaults to the current screen.
     * @param {Options} options - The options to merge with the current screen's options.
     *
     * @return {void}
     */
    mergeOptions(screen: AppScreenKeys, options: Options = {}): void {
        const currentScreen = screen || NavigationService.getCurrentScreen();
        Navigation.mergeOptions(currentScreen, options);
    },

    /**
     * Updates the props of a specific screen in the app.
     *
     * @template P - The type of the props being updated.
     * @param {AppScreenKeys} screen - The key of the screen to update the props for or current screen
     * @param {EnforcedProps<P>} props - The new props to apply to the screen.
     * @return {void}
     */
    updateProps<P extends object>(screen: AppScreenKeys, props: EnforcedProps<P>): void {
        const currentScreen = screen || NavigationService.getCurrentScreen();
        Navigation.updateProps(currentScreen, props);
    },

    /**
     * Updates the tabbar and re-renders the AppScreens.TabBar component.
     *
     * @return {void}
     */
    reRender(): void {
        // update the tabbar
        Object.keys(AppScreens.TabBar).forEach((tab) => {
            Navigation.mergeOptions(`bottomTab-${tab}`, {
                bottomTab: {
                    text: tab !== 'Actions' ? Localize.t(`global.${tab.toLowerCase()}`) : '',
                },
            });

            Navigation.updateProps(get(AppScreens.TabBar, tab), { timestamp: +new Date() });
        });
    },
};

export { Navigator };
