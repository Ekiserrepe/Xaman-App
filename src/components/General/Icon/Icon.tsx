/**
 * Icon
 *
    <Icon name={'IconLock'} size={25}></Icon>
 *
 */
import React from 'react';

import { Image, ImageStyle, StyleProp } from 'react-native';

import { AppSizes } from '@theme';
import { Images } from '@common/helpers/images';

/* Types ==================================================================== */
interface Props {
    name: Extract<keyof typeof Images, string>;
    size?: number;
    style?: StyleProp<ImageStyle>;
}

/* Component ==================================================================== */
const Icon: React.FC<Props> = ({ name, size = 25, style }) => (
    <Image
        resizeMode="contain"
        source={Images[name]}
        style={[{ width: AppSizes.moderateScale(size), height: AppSizes.moderateScale(size) }, style]}
    />
);

/* Export Component ==================================================================== */
export default Icon;
