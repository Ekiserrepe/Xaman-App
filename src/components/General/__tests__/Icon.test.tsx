import React from 'react';
import { Image } from 'react-native';
import renderer from 'react-test-renderer';

import { Images } from '@common/helpers/images';

import { Icon } from '../Icon';

describe('[Icon]', () => {
    it('renders correctly with default props', () => {
        const tree = renderer.create(<Icon name="IconFlaskConical" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with custom size', () => {
        const tree = renderer.create(<Icon name="IconFlaskConical" size={50} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with custom style', () => {
        const customStyle = { tintColor: 'red' };
        const tree = renderer.create(<Icon name="IconFlaskConical" style={customStyle} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('applies the correct image source', () => {
        const component = renderer.create(<Icon name="IconFlaskConical" />);
        const imageInstance = component.root.findByType(Image);
        expect(imageInstance.props.source).toEqual(Images.IconFlaskConical);
    });

    it('applies the correct size style', () => {
        const size = 50;
        const component = renderer.create(<Icon name="IconFlaskConical" size={size} />);
        const imageInstance = component.root.findByType(Image);
        expect(imageInstance.props.style).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    width: 78.57142857142857,
                    height: 78.57142857142857,
                }),
            ]),
        );
    });
});
