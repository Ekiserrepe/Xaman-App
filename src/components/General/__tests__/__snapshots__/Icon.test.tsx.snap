// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Icon] renders correctly with custom size 1`] = `
<Image
  resizeMode="contain"
  source={
    {
      "uri": "IconFlaskConical",
    }
  }
  style={
    [
      {
        "height": 78.57142857142857,
        "width": 78.57142857142857,
      },
      undefined,
    ]
  }
/>
`;

exports[`[Icon] renders correctly with custom style 1`] = `
<Image
  resizeMode="contain"
  source={
    {
      "uri": "IconFlaskConical",
    }
  }
  style={
    [
      {
        "height": 39.285714285714285,
        "width": 39.285714285714285,
      },
      {
        "tintColor": "red",
      },
    ]
  }
/>
`;

exports[`[Icon] renders correctly with default props 1`] = `
<Image
  resizeMode="contain"
  source={
    {
      "uri": "IconFlaskConical",
    }
  }
  style={
    [
      {
        "height": 39.285714285714285,
        "width": 39.285714285714285,
      },
      undefined,
    ]
  }
/>
`;
